Improve interrupt handling, set wifi lib's ISR, various improvements

kelnos opened the pull request:
Updates everything to build against ESP-IDF 4.2 and Rust 1.50 (which requires https://github.com/esp-rs/esp32-hal/pull/54).

I've implemented quite a few more things based on reading the ESP-IDF source to see what it does.  Unfortunately, the AP scan still does not return any results, and to make matters worse, now the app crashes with a `InstrProhibited` CPU exception shortly after the (non-functioning) AP scan finishes.

The main notable things that I did improve (in theory) was critical section/interrupt handling (which should now behave how ESP-IDF/FreeRTOS behaves), and allowing the WiFi blob to set and enable its ISR (which does end up getting run, though there are some oddities around `_intr_matrix_set()` that I don't understand).

I'm a bit disappointed that we still don't see any scan results.  I took the ESP-IDF WiFi scan example and stripped it down so it would more or less only call and do what we're doing, and it still works (with one exception; parsing scan results segfaults without calling `esp_supplicant_init()`).  So it's still unclear what ESP-IDF is doing that we aren't.

Locally I've started on a dummy impl of the WPA stuff, but it looks like we'll need to build and pull in the `wpa_supplicant` C code from ESP-IDF, as that's a mountain of code that I doubt we're going to replicate.  My initial thoughts were that maybe not having that code there was somehow causing the scan to fail, but I noted that even when I don't init the supplicant in the ESP-IDF sample app, the scan does find APs, even if it segfaults while parsing the data.

kelnos committed:
Regen, changes, and stubbing for ESP-IDF 4.2

kelnos committed:
Use TIMG0 from esp32_hal, not esp32

kelnos committed:
Bump esp32 dep to 0.10.0

kelnos committed:
Implement reentrant spinlock for phy critical sections

kelnos committed:
Load MAC address into calibration data

kelnos committed:
Implement block_time_tick properly for queue ops

kelnos committed:
Pass provided arg in _task_create_pinned_to_core()

kelnos committed:
Implement periodic timers

kelnos committed:
Implement interrupt handling

The esp-idf wifi driver expects to be able to set an ISR at runtime, not
compile-time. esp-idf & freertos allows this, but the rust xtensa
runtime support expects ISRs to be set up at compile-time.  To get
around this, we implement an empty ISR at compile-time, and call the
wifi driver's provided ISR after the driver tells us about it.

This isn't particularly ideal; this might mean it might be a good idea
to add first-class support for runtime ISR changes to the
xtensa-lx/esp32 crates.

Additionally, this implements _queue_send_from_isr() by just forwarding
to the non-ISR impl.  Which may not be safe.

kelnos committed:
Reenable ets_delay_us()

kelnos committed:
Implement intr_matrix_set()

This is a bit weird: if I just forward the impl to the built-in ROM
routine, we get some sort of livelock where the two cores keep sending
the same data back and forth in the same queue, but nothing happens (and
the wifi AP scan never completes).

The minimal (and probably not-quite-correct) implementation I have here
actually seems to wire up the interrupts correctly.  But of course we
still find zero APs, so...

kelnos committed:
Measure how long the scan takes

kelnos committed:
Print source string in strncpy()

kelnos committed:
Clean up the RF init code a bit

This also implements phy_update_wifi_mac_time()

kelnos committed:
Print event_base in system_event_handler

kelnos committed:
Move the ESP-IDF SpinLock stuff and extend it

Since we have to give the binary WiFi driver a SpinLock structure with
the memory layout it expects, we can't use safer Rust constructs in it.
Still, we can implement a spin lock in a similar way to how ESP-IDF
does, dropping to asm to use compare-and-set.

kelnos committed:
Add an InterruptManager

Something like this should probably go in the runtime or HAL, but we'll
get there later...

kelnos committed:
Use InterruptManager for critical sections & spinlocks

kelnos committed:
Get rid of ReentrantSpinLock

kelnos committed:
Implement strlen, strnlen, strncmp

kelnos committed:
Re-enable logging via wprintln

MabezDev commented:
For some reason my notifications for this repo were disabled, fortunately I found this by chance! This is awesome, and perfect timing as I am currently having another stab at this.

I've played around with the branch and it seems to work well on my board. I reached out to an espressif dev and he gave me a few troubleshooting tips; one being sniffing packets to see if the esp is emitting any, if it is then its likely the power & rf side is correctly configured.

I did manage to get a broadcast packet captured, sadly its malformed, but it does prove at the very least we have the phy & radio configured, heres the pcap: [esp32.pcapng.zip](https://github.com/esp-rs/esp32-wifi/files/6039059/esp32.pcapng.zip).

I'll keep playing and see if I can get anywhere.

kelnos commented:
Great!  Glad to hear you're looking back into this too.  It's been fun digging into this, but admittedly a little frustrating too.

Last night I was reading through `cpu_start.c` in esp-idf to see if maybe there was something (like a system timer) not getting intialized properly that the wifi implicitly depends on.  I was thinking about issues with power/rf, but wasn't sure how to test.  I think that's part of the problem; I'm just not sure how to debug a lot of these issues.

It's encouraging that you see something on the air, even if it's malformed.  At least that shows that there _is_ a problem we can tackle, even if we're not quite sure what it is yet.

I'm going to be out of town for a 4-day weekend, but will jump back in next week.

MabezDev commented:
> It's been fun digging into this, but admittedly a little frustrating too.

Oh I know that pain all too well :laughing:.

> It's encouraging that you see something on the air, even if it's malformed. At least that shows that there is a problem we can tackle, even if we're not quite sure what it is yet.

Indeed! I figured whilst I had my laptop configured to sniff, I would firstly try the master branch of this repo & also the idf scan example. Master produced nothing on the air at all, meaning this PR is a huge step in the right direction! The idf scan worked unsurprisingly, and here's the pcap: [esp32-idf.pcapng.zip](https://github.com/esp-rs/esp32-wifi/files/6039484/esp32-idf.pcapng.zip). Not sure how useful that will be, other than confirming the phy & rf is setup properly (the 802.11 radio information section is the same in both).

kelnos commented:
Strangely I can't seem to get my laptop to sniff wifi frames.  It claims to support monitor mode, but I when I set things up I get zero packets (not even any from wifi devices that function properly).  Something to dig into another day...

Speaking of power, I noticed a spot in esp-idf `rtc_init()` where it sets a register to force the wifi chip to power up.  I'm not aware of anything we're doing that would also do that, unless I've missed it.

MabezDev commented:
> Strangely I can't seem to get my laptop to sniff wifi frames. It claims to support monitor mode, but I when I set things up I get zero packets (not even any from wifi devices that function properly). Something to dig into another day...

If you're running linux, `iw` didn't work for me, instead a friend directed to me towards `aircrack-ng` which did work.

> Speaking of power, I noticed a spot in esp-idf rtc_init() where it sets a register to force the wifi chip to power up. I'm not aware of anything we're doing that would also do that, unless I've missed it.

I think by default, most things are enabled, and have to be powered down; though I'm not sure that's true for all cases. I think in this case you mean `RTC_CNTL_WIFI_FORCE_PU`?

I think this snippet from line 1368 in `rtc_cntl_reg.h` shows that by default, the bit is set.
```c
/* RTC_CNTL_WIFI_FORCE_PU : R/W ;bitpos:[18] ;default: 1'd1 ; */
/*description: wifi force power up*/
#define RTC_CNTL_WIFI_FORCE_PU  (BIT(18))
#define RTC_CNTL_WIFI_FORCE_PU_M  (BIT(18))
#define RTC_CNTL_WIFI_FORCE_PU_V  0x1
#define RTC_CNTL_WIFI_FORCE_PU_S  18
```

MabezDev commented:
I've still been trying to get this working, no luck so far.

> though there are some oddities around _intr_matrix_set() that I don't understand

I believe you have actually set this up correctly now, using the `esp32_hal`. I think calling `core::mem::transmute::<_, unsafe extern "C" fn(c_int, c_uint, c_uint)>(0x4000681c)(cpu_no, model_num, intr_num);` will only map the interrupt, you would also need to set a handler & enable the interrupt on the CPU for this method to work. I.e also call `xt_set_interrupt_handler` & `xt_ints_on`.

I think I'm going to look into the allocation next, seem strange that the esp can consistently output 802.11 Probe Request section, yet the Wireless Management section is garbage. Perhaps the alloc needs a specific alignment, or its just plain wrong. Though this is clutching at straws really.

kelnos commented:
My assumption was that `_set_isr()` sets the handler and `_ints_in()` enable the interrupt (both of which I implemented in terms of our rust-based runtime stuff).  Since in an esp-idf app the `_intr_matrix_set()` call goes right to the ROM, it's weird to me that doing so here doesn't do the right thing.  It makes me think that we're still missing something when it comes to interrupts.

kelnos commented:
Allocation alignment seems like a good idea to look into.  I'm pretty low on ideas as well...

One potential thing I'm worried about is deadlocks with the queue code since they all need to take a global lock to do anything with a queue.  I've been trying to remove all the `BTree` holding those and just return pointers (using a technique similar to the spinlocks where I use `Box::leak()` and `Box::from_raw()`).  ~However, I've been getting odd behavior; in the function that creates a queue, I return what looks like a reasonable pointer, but then when the wifi lib tries to use it in the send/recv functions, I get a very strange, unaligned pointer (which then causes a CPU exception).~  Nevermind, I actually did figure this out a couple weeks ago.

I still have some more to do to debug this, just haven't had the time this week.

kelnos committed:
Simplify Queue; remove global queue table

kelnos commented:
Still can't seem to get radio capture working.  I did manage to get aircrack-ng to run, but it just sits there and never sees anything.  Tried on two different laptops, but they both have Atheros/Qualcomm chipsets... I think I have another laptop with a Broadcom chipset (tho they tend to be supported less), will give that a try later.

kelnos commented:
Looking at the allocator as well.  ESP-IDF defines the three like this:
```
malloc() -> heap_caps_malloc_default() -> { heap_caps_malloc(DEFAULT | INTERNAL) || heap_caps_malloc(DEFAULT) }
wifi_malloc() -> malloc()
malloc_internal() -> heap_caps_malloc(8BIT | DMA | INTERNAL)
```
Our setup is to use DRAM for `malloc_internal()` and the "general" allocator for `wifi_malloc()` (which I _think_ in the default setup will also just use DRAM, but can also use IRAM as well).

I did try using the DRAM allocator for everything, but that didn't change anything.  Also tried using `DRAM_ALLOCATOR.allocate()` directly; rather than using `RawVec`, but also no change.

kelnos commented:
Hmm, I looked at `_task_create_pinned_to_core()`. It looks like the wifi blob calls it with a core ID of 0, which I believe is the main/Pro core?  And yet we're starting that task on the App core... could that be an issue?

MabezDev commented:
> Hmm, I looked at _task_create_pinned_to_core(). It looks like the wifi blob calls it with a core ID of 0, which I believe is the main/Pro core? And yet we're starting that task on the App core... could that be an issue?

Good spot! I tried setting `wifi_task_core_id` to `1` in the `wifi_init_config_t` but the scan just times out, will need to explore that a bit further.

kelnos commented:
Ah, damn, yeah, I see that too.  That _is_ supposed to work, it seems; I set `CONFIG_ESP32_WIFI_TASK_PINNED_TO_CORE_1` in the esp-idf scan example, and it did work...  Very strange that we see a timeout...

kelnos commented:
EDIT: Nevermind.  It looks like this isn't binary tracing; it's just the esp-idf monitor command regexing for `0x4[0-9a-f]{7}` in each line it encounters, and then looking in the binary for a function name and printing it.  It's not some tracing that prints when that function was actually run.

Another thing I noticed a while ago but am thinking more about again: it appears that the binary wifi lib does have some minimal function tracing enabled.  I've been using the esp-idf `idf.py monitor` program to view output (for some reason `screen` isn't showing line breaks properly), and I see some of these trace messages (interestingly they show up in orange rather than my terminal's regular green).  There are a few trace messages that show up in the esp-idf scan app as well, but completely different messages show up in our test app.

Fairly early on, I see:
```
WIFI core APP: _log_write(3, wifi, %c (%d) %s:, 1)
WIFI core APP: _log_writev(3, wifi, Init static rx buffer num: %d, ...)
WIFI core APP: _log_write(3, wifi, %s, 3f4168dc)
WIFI core APP: _log_write(3, wifi, %c (%d) %s:, 3f410b30)
WIFI core APP: _log_writev(3, wifi, Init dynamic rx buffer num: %d, ...)
WIFI core APP: _log_write(3, wifi, %s, 3f4168fc)
WIFI core APP: _timer_setfn(0x3ffb2e54, 0x40104900, 0x0)
0x40104900: lmac_stop_hw_txq at ??:?

WIFI core APP: _timer_setfn(0x3ffb2e68, 0x401048f0, 0x0)
0x401048f0: lmac_stop_hw_txq at ??:?

WIFI core APP: _timer_setfn(0x3ffb2e7c, 0x401048e0, 0x0)
0x401048e0: lmac_stop_hw_txq at ??:?

WIFI core APP: _timer_setfn(0x3ffb2eb8, 0x401048d0, 0x0)
0x401048d0: lmac_stop_hw_txq at ??:?

WIFI core APP: _timer_setfn(0x3ffb2e40, 0x401048c0, 0x0)
0x401048c0: lmac_stop_hw_txq at ??:?

WIFI core APP: _timer_setfn(0x3ffb2e90, 0x401048b0, 0x0)
0x401048b0: lmac_stop_hw_txq at ??:?

WIFI core APP: _timer_setfn(0x3ffb2ea4, 0x401048a0, 0x0)
0x401048a0: lmac_stop_hw_txq at ??:?

WIFI core APP: _timer_setfn(0x3ffb290c, 0x400f9e40, 0x0)
0x400f9e40: ieee80211_rfid_locp_recv at ??:?

WIFI core APP: _timer_setfn(0x3ffb2920, 0x400f9e30, 0x0)
0x400f9e30: ieee80211_rfid_locp_recv at ??:?
```

The first series looks concerning; why is the driver stopping the transmit queue?

If I then scroll down way farther in the output, I see:

```
WIFI core APP: phy_printf(phy_version: %d, %s, %s, %s, %d, %d)
[snip 26 lines]
WIFI core APP: _timer_setfn(0x3ffb2aa4, 0x400fddd0, 0x0)
0x400fddd0: ieee80211_timer_process at ??:?

WIFI core APP: _timer_setfn(0x3ffb2ab8, 0x400fddd0, 0x1)
0x400fddd0: ieee80211_timer_process at ??:?

WIFI core APP: intr_matrix_set(0,0,0)
WIFI core APP: _set_isr(0, 400881cc, 0)
WIFI core APP: _ints_on(1)
WIFI core APP: _timer_arm(3ffb2e40, 256, false)
WIFI core APP: _wifi_zalloc(552)
WIFI core APP: alloc(552, false, true) -> 3ffb94b8
WIFI core APP: _wifi_zalloc(760)
WIFI core APP: alloc(760, false, true) -> 3ffb96e0
WIFI core APP: _wifi_zalloc(212)
WIFI core APP: alloc(212, false, true) -> 3ffb99d8
WIFI core APP: _free(0x3ffb9aac)
WIFI core APP: _mutex_lock (0x3ffb3df0)
WIFI core APP: _mutex_unlock (0x3ffb3df0)
WIFI core APP: _timer_setfn(0x3ffb22c8, 0x400ee868, 0x0)
0x400ee868: wifi_softap_stop at ??:?

WIFI core APP: _log_write(3, wifi, %c (%d) %s:, 3ffb22c8)
WIFI core APP: _log_writev(3, wifi, mode : sta (%02x:%02x:%02x:%02x:%02x:%02x), ...)
WIFI core APP: _log_write(3, wifi, %s, 3f410878)
WIFI core APP: __popcountsi2(1) -> 1
WIFI core APP: system_event_handler(WIFI_EVENT, 2, 0, 0, 4294967295)
WIFI core APP: WARNING: phy_update_country_info(CN) unimplemented
WIFI core APP: _semphr_give(3ffb4520)
WIFI core PRO: _semphr_take -> 1
WIFI core APP: _phy_common_clock_disable
WIFI core PRO: _free(0x3ffb93f8)
WIFI core APP: _queue_recv(0x3ffb4500, ax3fffdfa8, ffffffff)
WIFI core PRO: _mutex_lock (0x3ffb3d80)
WIFI core PRO: _mutex_unlock (0x3ffb3d80)
WIFI core PRO: _mutex_lock (0x3ffb3df0)
WIFI core PRO: _mutex_unlock (0x3ffb3df0)
WIFI core PRO: _mutex_lock (0x3ffb3df0)
WIFI core PRO: _mutex_unlock (0x3ffb3df0)
WIFI core PRO: _wifi_zalloc(20)
WIFI core PRO: alloc(20, false, true) -> 3ffb9ab8
WIFI core PRO: _wifi_thread_semphr_get -> 3ffb4520
WIFI core PRO: _mutex_lock (0x3ffb3d80)
WIFI core PRO: _wifi_int_disable(3ffb3de8)
WIFI core PRO: _wifi_int_restore(3ffb3de8, 0)
WIFI core PRO: _task_ms_to_tick(10) -> 800000
WIFI core PRO: _queue_send(0x3ffb4500, 0x3ffffce8, 800000)
WIFI core PRO: _queue_send 00000006 3ffb9ab8 8 -> TRUE
WIFI core PRO: _queue_send -> 1
WIFI core APP: _queue_recv 00000006 3ffb9ab8 8 -> TRUE
WIFI core PRO: _mutex_unlock (0x3ffb3d80)
WIFI core APP: _queue_recv -> 1
WIFI core PRO: _semphr_take(3ffb4520, 4294967295)
WIFI core APP: _wifi_int_disable(3ffb3de8)
WIFI core APP: _wifi_int_restore(3ffb3de8, 0)
WIFI core APP: _phy_common_clock_enable
WIFI core APP: _timer_disarm(3ffb94bc)
WIFI core APP: _timer_done(0x3ffb94bc) 
WIFI core APP: _timer_setfn(0x3ffb94bc, 0x400fea90, 0x0)
0x400fea90: cnx_connect_timeout at ??:?
```

`cnx_connect_timeout` definitely looks concerning.  What is "cnx"?  `wifi_softap_stop` is odd, as we set `WIFI_MODE_STA` in the test app.

kelnos committed:
Implement semphr_take's block_time_tick properly

kelnos committed:
Implement hacky varargs sprintf and logging in C

There's still something wrong with how rustc is emitting C vararg
instructions, so I've moved the implementations to C for now.

kelnos committed:
Disable interrupts while logging

kelnos commented:
I'm starting to think that timing is playing a factor here too.  With nearly all the debug logging disabled (and the new varargs stuff enabled that I just pushed), I managed to trigger a new error:
```
E (0) wifi:wpa michael mic failure cb empty
```
I was only able to repro it a few times last night, and then today I made some changes (can't remember what :( ) and I'm no longer able to repro it.  I was really excited to see it, because I _think_ that might only happen when the driver is trying to parse some WPA-related stuff in a received wifi packet (but fails because we haven't set up any of the WPA stuff yet).  Really bummed I can't seem to get back to the place where I was seeing that error, even if I couldn't repro it consistently.

But doing other "silly" things can change the overall result: disabling a lot of the more frequent logging (will push a logging refactor commit shortly) will often stop the `InstrProhibited` message.  Sometimes I get a timeout error from the scan operation instead of returning 0 APs.  Prior to all this, I was nearly 100% consistently getting 0 APs with an `InstrProhibited` shortly after it.

I've also been looking into the locking code... it's definitely possible we should be disabling interrupts more often, like when dealing with the global locks that protect the mutexes, semaphores, etc., but possibly other places as well.  I feel like we're at worst getting deadlocks at times, and at best creating enough contention that the timing gets weird.  At one point I did something that started causing a deadlock in `queue_send()`, and just changing the logging in that function "fixed" the deadlock.  (The ironic thing is that rust usually protects us from data races, but we're using mostly `unsafe` rust so far...)

I also tried making `ets_delay_us()` call the ROM function rather than the rust code we have, and that _might_ have been related to triggering the "michael mic" error message.  Certainly logging on each of those delay calls (which I ended up disabling) is causing the delay to be longer than the delay the caller is looking for.  Maybe it matters?  I guess usually those kinds of delays are put into place to allow hardware to settle, so a _longer_ delay usually isn't a problem, but who knows...

kelnos committed:
Refactor logging

I'm worried about too-frequent logging throwing off timing to a degree
that it's playing a role in the issues we're having, especially when we
log in a function like ets_delay_us().

This adds (compile-time) configurable log levels with macros for logging
at each level.

kelnos committed:
Call into ROM for ets_delay_us()

No strong reason for this, but it's provided by the ESP32 ROM so we
might as well use it.

kelnos committed:
Disable interrupts when working with queues

This should hopefully make it unnecessary to have separate impls for
queue_send() and queue_send_from_isr(), and ensure that regular code and
interrupt code can't step on each other when touching queues (since it
seems the WiFi lib's ISR makes use of the queues).

kelnos committed:
Fix strncpy() implementation

Previously, if a NUL-initialized buffer was passed as the destination,
only NULs would get written to the entire buffer.

kelnos committed:
Use 5 instead of 7 for the DPORT interrupt disable level

Not sure why 7 was chosen here, so there might be an actual reason.  But
esp-idf uses 5 for this by default.

kelnos committed:
Fix _read_mac MAC modifications and return value

AdityaHPatwardhan commented:
Hi @kelnos I was trying out this PR, First of all Nice work indeed  👏. 
 I have got this PR to build by providing command 
`cargo espflash --release --tool=cargo --chip=esp32 --example=test /dev/cu.SLAB_USBtoUART` ( please correct me if I am wrong) . Then I used picocom for the output. But the logs are very short there.

however, I noticed you have shared logs that you obtained from `idf.py monitor`. Can you please share some details on how you managed that and also the required steps. I would like to debug this issue on my side and maybe help you solve it.

kelnos commented:
@AdityaHPatwardhan it's fortunately pretty easy to use the monitor app from esp-idf.  Just run:
```
$ESP_IDF_SRCDIR/tools/idf_monitor.py --port /dev/ttyUSB0 --baud 921600 target/xtensa-esp32-none-elf/release/examples/test
```
From the root of the esp32-wifi directory.  Of course replace `$ESP_IDF_SRCDIR` with wherever your esp-idf source checkout lives, and `/dev/ttyUSB0` with your esp32 board's serial device path.

AdityaHPatwardhan commented:
okay, thanks 👍

kelnos commented:
@AdityaHPatwardhan also I just wrote released [cargo-espmonitor](https://github.com/kelnos/espmonitor).  You can install it with `cargo install cargo-espmonitor`. :)

AdityaHPatwardhan commented:
@kelnos the `cargo-espmonitor` looks very nice 👌.
Meanwhile on this PR.
1) I tried to compare the assembly execution of some important wifi functions with that of the code built in the `esp-idf` but that seems in order.
2) I tried to sniff packets on the latest commit but couldn't find anything. But when I tried sniffing for 339ea23710d2472b8104bbcaef17ea791fe28017, I got the malformed packets as mentioned earlier by @MabezDev 
The malformed packets might have been caused by improper wifi calibration or could be memory corruption at some point.
I haven't got a clear direction from this as of yet, but will try to debug further.

kelnos commented:
Ah, damn, so maybe I broke something after https://github.com/esp-rs/esp32-wifi/commit/339ea23710d2472b8104bbcaef17ea791fe28017 ... unfortunately my laptop's wifi adapter doesn't properly support monitor mode, so I can't test this.  If you have time, could you bisect and see which commit broke it?

shubhamkulkarni97 commented:
Hi @kelnos @MabezDev,

I have some questions regarding implementation of OS abstraction layer for Wi-Fi driver.
In [task_create_wrapper](https://github.com/esp-rs/esp32-wifi/blob/dcb50b7d45627961598e35dec32a7b13e2f12ab8/src/compatibility/osi.rs#L620), app core is enabled instead of creating a task/thread. As a task is not created in wrapper, stack will not be allocated for storing local variables. How is this case handle in current implementation?

ESP Wi-Fi driver assumes that a task is created when `task_create_wrapper` returns success.

kelnos commented:
@shubhamkulkarni97 I'm not responsible for that initial code, but my assumption is that @arjanmels was just hoping that the wifi driver would only create a single task so he wouldn't need to write a full preemptive task scheduler.

The task that does end up running after that does appear to run fine -- at least, we see log messages appearing on the other core from the code pointed to by `task_func` running on it -- so I don't think what you're suggesting is actually a problem.

If you dig into the `start_app_core()` function in esp32-hal, you'll see that it does indeed properly initialize the second core, including setting up a stack.

arjanmels commented:
@kelnos your assumption is correct. To simplify matters I indeed simply dedicated a core for WiFi. Great that your are picking this up. Still seems a hard but to crack.

kelnos commented:
Had an old Mac Mini lying around (with Linux on it), and it appears monitor mode works on its wifi, yay!  @AdityaHPatwardhan I tried with the current tip of this branch (dcb50b7d45627961598e35dec32a7b13e2f12ab8) and I do see (malformed) packets from the ESP32, so not sure what's going on with your setup :(

AdityaHPatwardhan commented:
@kelnos I checked it again, I am getting the malformed packets on the latest commit as well. I must have missed something, or did not capture the packet clearly 🤦.  Sorry !!

kelnos committed:
Bump deps to latest

Sadly this breaks things; execution hangs in TimerFactoryImpl::new()
somewhere.

MabezDev commented:
We've just published [esp-rs/esp-wifi](https://github.com/esp-rs/esp-wifi) which currently supports bare metal WiFi on the esp32 & esp32c3, therefore we will now be archiving this repo.

Thanks for all the hard work folks!
